File: implementations/android-sdk/uitests/src/main/kotlin/com/contentful/optimization/uitests/tests/IdentifiedVariantsTests.kt
package com.contentful.optimization.uitests.tests

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.By
import androidx.test.uiautomator.UiDevice
import androidx.test.uiautomator.UiScrollable
import androidx.test.uiautomator.UiSelector
import com.contentful.optimization.uitests.support.AppLauncher
import com.contentful.optimization.uitests.support.TestHelpers
import com.contentful.optimization.uitests.support.clearProfileState
import org.junit.Assert
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class IdentifiedVariantsTests {
    private lateinit var device: UiDevice

    companion object {
        @JvmStatic
        @BeforeClass
        fun setUpClass() {
            val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())

            // Step 1: launch and clear any leftover profile state.
            AppLauncher.launchApp(device)
            clearProfileState(device)

            // Step 2: wait for identify-button, then tap it.
            TestHelpers.waitForElement(device, By.res("identify-button"), TestHelpers.ELEMENT_TIMEOUT)
            TestHelpers.waitAndTap(device, By.res("identify-button"))

            // Step 3: wait for reset-button, confirming identify succeeded.
            TestHelpers.waitForElement(device, By.res("reset-button"), TestHelpers.EXTENDED_TIMEOUT)

            // Step 4: terminate the app.
            AppLauncher.forceStop(device)

            // Step 5: relaunch as a new instance so identified state is rehydrated
            // from persistent storage.
            AppLauncher.launchApp(device)

            // Step 6: wait for reset-button in the relaunched app. This proves
            // (a) the relaunch finished loading and (b) the identified profile
            // survived the cold start — the precondition every test in this suite
            // needs.
            TestHelpers.waitForElement(device, By.res("reset-button"), TestHelpers.EXTENDED_TIMEOUT)
        }
    }

    @Before
    fun setUp() {
        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
    }

    // MARK: - common variants

    @Test
    fun testShouldDisplayMergeTagContentWithResolvedValue() {
        val expected = "This is a merge tag content entry that displays the visitor's continent \"EU\" embedded within the text. [Entry: 1MwiFl4z7gkwqGYdvCmr8c]"
        // The Android app resolves merge tags asynchronously, so wait for the
        // resolved description rather than asserting on it immediately.
        TestHelpers.waitForElement(device, By.desc(expected), TestHelpers.ELEMENT_TIMEOUT)
    }

    @Test
    fun testShouldDisplayVariantForVisitorsFromEurope() {
        TestHelpers.waitForElement(device, By.res("entry-text-4ib0hsHWoSOnCVdDkizE8d"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a variant content entry for visitors from Europe. [Entry: 4ib0hsHWoSOnCVdDkizE8d]"
        Assert.assertNotNull(
            "Expected Europe continent variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    @Test
    fun testShouldDisplayVariantForDesktopBrowserVisitors() {
        TestHelpers.waitForElement(device, By.res("entry-text-xFwgG3oNaOcjzWiGe4vXo"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a variant content entry for visitors using a desktop browser. [Entry: xFwgG3oNaOcjzWiGe4vXo]"
        Assert.assertNotNull(
            "Expected desktop browser variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    // MARK: - identified user variants

    @Test
    fun testShouldDisplayVariantForReturnVisitors() {
        TestHelpers.waitForElement(device, By.res("entry-text-2Z2WLOx07InSewC3LUB3eX"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a variant content entry for return visitors. [Entry: 2Z2WLOx07InSewC3LUB3eX]"
        Assert.assertNotNull(
            "Expected return visitor variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    @Test
    fun testShouldDisplayVariantBForABCExperiment() {
        TestHelpers.waitForElement(device, By.res("entry-text-5XHssysWUDECHzKLzoIsg1"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a variant content entry for an A/B/C experiment: B [Entry: 5XHssysWUDECHzKLzoIsg1]"
        Assert.assertNotNull(
            "Expected A/B/C experiment variant B label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    @Test
    fun testShouldDisplayVariantForVisitorsWithCustomEvent() {
        TestHelpers.waitForElement(device, By.res("entry-text-6zqoWXyiSrf0ja7I2WGtYj"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a variant content entry for visitors with a custom event. [Entry: 6zqoWXyiSrf0ja7I2WGtYj]"
        Assert.assertNotNull(
            "Expected custom event variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    @Test
    fun testShouldDisplayVariantForIdentifiedUsers() {
        scrollTo("entry-text-7pa5bOx8Z9NmNcr7mISvD")
        TestHelpers.waitForElement(device, By.res("entry-text-7pa5bOx8Z9NmNcr7mISvD"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a variant content entry for identified users. [Entry: 7pa5bOx8Z9NmNcr7mISvD]"
        Assert.assertNotNull(
            "Expected identified users variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    // MARK: - nested optimization variants

    @Test
    fun testShouldDisplayLevel0NestedVariantForReturnVisitors() {
        scrollTo("entry-text-2KIWllNZJT205BwOSkMINg")
        TestHelpers.waitForElement(device, By.res("entry-text-2KIWllNZJT205BwOSkMINg"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a level 0 nested variant entry. [Entry: 2KIWllNZJT205BwOSkMINg]"
        Assert.assertNotNull(
            "Expected level 0 nested variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    @Test
    fun testShouldDisplayLevel1NestedVariantForReturnVisitors() {
        scrollTo("entry-text-5a8ONfBdanJtlJ39WWnH1w")
        TestHelpers.waitForElement(device, By.res("entry-text-5a8ONfBdanJtlJ39WWnH1w"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a level 1 nested variant entry. [Entry: 5a8ONfBdanJtlJ39WWnH1w]"
        Assert.assertNotNull(
            "Expected level 1 nested variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    @Test
    fun testShouldDisplayLevel2NestedVariantForReturnVisitors() {
        scrollTo("entry-text-4hDiXxYEFrXHXcQgmdL9Uv")
        TestHelpers.waitForElement(device, By.res("entry-text-4hDiXxYEFrXHXcQgmdL9Uv"), TestHelpers.ELEMENT_TIMEOUT)
        val expected = "This is a level 2 nested variant entry. [Entry: 4hDiXxYEFrXHXcQgmdL9Uv]"
        Assert.assertNotNull(
            "Expected level 2 nested variant label to be visible",
            device.findObject(By.desc(expected)),
        )
    }

    private fun scrollTo(resourceId: String) {
        try {
            UiScrollable(UiSelector().resourceId("main-scroll-view"))
                .apply { setMaxSearchSwipes(10) }
                .scrollIntoView(UiSelector().resourceId(resourceId))
        } catch (_: Exception) {
            // Element may already be visible or scrolling not possible
        }
    }
}

File: implementations/ios-sdk/uitests/Tests/IdentifiedVariantsTests.swift
import XCTest

/// 1:1 port of `displays-identified-user-variants.test.js`.
///
/// Verifies that once a visitor has been identified and the app has been
/// relaunched, the SDK resolves and renders the correct variant for each
/// optimized entry on screen. Covers common variants (merge tag, continent,
/// device), identified-user-only variants (return visitor, A/B/C bucket,
/// custom-event audience, identified audience), and nested optimization
/// variants across three levels of depth.
final class IdentifiedVariantsTests: XCTestCase {
    let app = XCUIApplication()

    override class func setUp() {
        let app = XCUIApplication()
        app.launch()
        clearProfileState(app: app)

        waitForElement(app.buttons["identify-button"], timeout: ELEMENT_VISIBILITY_TIMEOUT)
        app.buttons["identify-button"].tap()
        waitForElement(app.buttons["reset-button"], timeout: ELEMENT_VISIBILITY_TIMEOUT)

        // Relaunch as a new instance so the identified state is rehydrated from
        // persistent storage. The relaunched app derives the identify/reset
        // control from the rehydrated identified profile, so `reset-button`
        // (not `identify-button`) appears. Waiting for it both confirms the
        // relaunch finished loading and proves the identified profile actually
        // survived the cold start — the precondition every test in this suite
        // needs.
        app.terminate()
        app.launch()
        waitForElement(app.buttons["reset-button"], timeout: ELEMENT_VISIBILITY_TIMEOUT)
    }

    // MARK: - common variants

    func testShouldDisplayMergeTagContentWithResolvedValue() {
        let element = app.otherElements["entry-text-1MwiFl4z7gkwqGYdvCmr8c"]
        waitForElement(element, timeout: ELEMENT_VISIBILITY_TIMEOUT)

        // Asserted against the element's label (not via a label subscript) because
        // the merge tag label exceeds XCUITest's 128-character identifier limit.
        let expected = "This is a merge tag content entry that displays the visitor's continent \"EU\" embedded within the text. [Entry: 1MwiFl4z7gkwqGYdvCmr8c]"
        XCTAssertEqual(element.label, expected, "Expected merge tag variant label to be visible")
    }

    func testShouldDisplayVariantForVisitorsFromEurope() {
        waitForElement(app.otherElements["entry-text-4ib0hsHWoSOnCVdDkizE8d"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a variant content entry for visitors from Europe. [Entry: 4ib0hsHWoSOnCVdDkizE8d]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected Europe continent variant label to be visible")
    }

    func testShouldDisplayVariantForDesktopBrowserVisitors() {
        waitForElement(app.otherElements["entry-text-xFwgG3oNaOcjzWiGe4vXo"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a variant content entry for visitors using a desktop browser. [Entry: xFwgG3oNaOcjzWiGe4vXo]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected desktop browser variant label to be visible")
    }

    // MARK: - identified user variants

    func testShouldDisplayVariantForReturnVisitors() {
        waitForElement(app.otherElements["entry-text-2Z2WLOx07InSewC3LUB3eX"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a variant content entry for return visitors. [Entry: 2Z2WLOx07InSewC3LUB3eX]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected return visitor variant label to be visible")
    }

    func testShouldDisplayVariantBForABCExperiment() {
        waitForElement(app.otherElements["entry-text-5XHssysWUDECHzKLzoIsg1"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a variant content entry for an A/B/C experiment: B [Entry: 5XHssysWUDECHzKLzoIsg1]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected A/B/C experiment variant B label to be visible")
    }

    func testShouldDisplayVariantForVisitorsWithCustomEvent() {
        waitForElement(app.otherElements["entry-text-6zqoWXyiSrf0ja7I2WGtYj"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a variant content entry for visitors with a custom event. [Entry: 6zqoWXyiSrf0ja7I2WGtYj]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected custom event variant label to be visible")
    }

    func testShouldDisplayVariantForIdentifiedUsers() {
        waitForElement(app.otherElements["entry-text-7pa5bOx8Z9NmNcr7mISvD"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a variant content entry for identified users. [Entry: 7pa5bOx8Z9NmNcr7mISvD]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected identified users variant label to be visible")
    }

    // MARK: - nested optimization variants

    func testShouldDisplayLevel0NestedVariantForReturnVisitors() {
        waitForElement(app.otherElements["entry-text-2KIWllNZJT205BwOSkMINg"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a level 0 nested variant entry. [Entry: 2KIWllNZJT205BwOSkMINg]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected level 0 nested variant label to be visible")
    }

    func testShouldDisplayLevel1NestedVariantForReturnVisitors() {
        waitForElement(app.otherElements["entry-text-5a8ONfBdanJtlJ39WWnH1w"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a level 1 nested variant entry. [Entry: 5a8ONfBdanJtlJ39WWnH1w]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected level 1 nested variant label to be visible")
    }

    func testShouldDisplayLevel2NestedVariantForReturnVisitors() {
        waitForElement(app.otherElements["entry-text-4hDiXxYEFrXHXcQgmdL9Uv"],
                       timeout: ELEMENT_VISIBILITY_TIMEOUT)

        let expected = "This is a level 2 nested variant entry. [Entry: 4hDiXxYEFrXHXcQgmdL9Uv]"
        XCTAssertTrue(app.otherElements[expected].waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Expected level 2 nested variant label to be visible")
    }
}
